src/dbnode/persist/fs/merger.go

```diff
@@ -0,0 +1,293 @@
+		// because the series gets persisted each loop, so previous iterations'
+		// reader and iterator will never be needed.
+		segReader = srPool.Get()
+		multiIter = multiIterPool.Get()
+		// Initialize this here with nil to be reset before each iteration's
+		// use.
+		sliceOfSlices = xio.NewReaderSliceOfSlicesFromBlockReadersIterator(nil)
+		// Reused context for use in mergeWith.Read, since they all do a
+		// BlockingClose after usage.
+		ctx = context.NewContext()
+
+		// We keep track of IDs/tags to finalize at the end of merging. This
+		// only applies to those that come from disk Reads, since the whole
+		// lifecycle of those IDs/tags are contained to this function. We don't
+		// want finalize the IDs from memory since other components may have
+		// ownership over it.
+		//
+		// We must only finalize these at the end of this function, since the
+		// flush preparer's underlying writer holds on to those references
+		// until it is closed (closing the PreparedDataPersist at the end of
+		// this merge closes the underlying writer).
+		idsToFinalize  = make([]ident.ID, 0, reader.Entries())
+		tagsToFinalize = make([]ident.Tags, 0, reader.Entries())
+	)
+	defer func() {
+		segReader.Finalize()
+		multiIter.Close()
+		for _, res := range idsToFinalize {
+			res.Finalize()
+		}
+		for _, res := range tagsToFinalize {
+			res.Finalize()
+		}
+	}()
+
+	// The merge is performed in two stages. The first stage is to loop through
+	// series on disk and merge it with what's in the merge target. Looping
+	// through disk in the first stage is done intentionally to read disk
+	// sequentially to optimize for spinning disk access. The second stage is to
+	// persist the rest of the series in the merge target that were not
+	// persisted in the first stage.
+
+	// First stage: loop through series on disk.
+	for id, tagsIter, data, _, err := reader.Read(); err != io.EOF; id, tagsIter, data, _, err = reader.Read() {
+		idsToFinalize = append(idsToFinalize, id)
+		if err != nil {
+			return err
+		}
+
+		// Reset BlockReaders.
+		brs = brs[:0]
+		br = br[:0]
+		br = append(br, blockReaderFromData(data, segReader, startTime, blockSize))
+
+		// Check if this series is in memory (and thus requires merging).
+		mergeWithData, hasData, err := mergeWith.Read(ctx, id, blockStart, nsCtx)
+		if err != nil {
+			return err
+		}
+		if hasData {
+			br = append(br, mergeWithData...)
+		}
+		brs = append(brs, br)
+
+		sliceOfSlices.Reset(brs)
+		multiIter.ResetSliceOfSlices(sliceOfSlices, nsCtx.Schema)
+
+		// tagsIter is never nil.
+		tags, err := convert.TagsFromTagsIter(id, tagsIter, identPool)
+		tagsIter.Close()
+		tagsToFinalize = append(tagsToFinalize, tags)
+		if err != nil {
+			return err
+		}
+		if err := persistIter(prepared.Persist, multiIter, startTime,
+			id, tags, blockAllocSize, nsCtx.Schema, encoderPool); err != nil {
+			return err
+		}
+	}
+	// Second stage: loop through rest of the merge target that was not captured
```

Comment:
"loop through any series in the merge target that were not captured in the first stage"